File: sdk/md2html_client.py
"""Small, dependency-free Python client for the MD2HTML API.

The public API is intentionally tiny: convert Markdown, convert batches, create
an API key, inspect usage, pretty-print JSON, and calculate text statistics.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Dict, Iterable, List, Optional, Union

__version__ = "0.1.0"
DEFAULT_BASE_URL = "http://147.15.103.217/md2html"


class Md2HTMLAPIError(RuntimeError):
    """Raised when the API or the HTTP transport returns an error."""

    def __init__(
        self,
        message: str,
        *,
        status_code: int = 0,
        url: str = "",
        response_body: str = "",
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.url = url
        self.response_body = response_body


class Md2HTMLClient:
    """Client for the MD2HTML HTTP API.

    Args:
        base_url: API root. Defaults to the hosted MD2HTML service.
        api_key: Optional ``mk_...`` key sent as ``X-API-Key``.
        timeout: Per-request timeout in seconds.
    """

    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        api_key: Optional[str] = None,
        timeout: float = 30.0,
    ) -> None:
        if not isinstance(base_url, str) or not base_url.strip():
            raise ValueError("base_url must be a non-empty URL")
        if timeout <= 0:
            raise ValueError("timeout must be greater than zero")
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key or ""
        self.timeout = float(timeout)
        self.last_response: Optional[Any] = None

    def _request(
        self,
        method: str,
        path: str,
        payload: Optional[Dict[str, Any]] = None,
    ) -> Any:
        """Send one JSON request and return its decoded response body."""
        url = self.base_url + "/" + path.lstrip("/")
        headers = {
            "Accept": "application/json",
            "User-Agent": "md2html-python-sdk/" + __version__,
        }
        body = None
        if payload is not None:
            body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
            headers["Content-Type"] = "application/json"
        if self.api_key:
            headers["X-API-Key"] = self.api_key

        request = urllib.request.Request(
            url, data=body, headers=headers, method=method.upper()
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                status = response.status
                raw = response.read()
        except urllib.error.HTTPError as error:
            raw = error.read()
            text = raw.decode("utf-8", errors="replace")
            raise Md2HTMLAPIError(
                "MD2HTML API returned HTTP {}".format(error.code),
                status_code=error.code,
                url=url,
                response_body=text,
            ) from error
        except urllib.error.URLError as error:
            raise Md2HTMLAPIError(
                "Could not reach MD2HTML API: {}".format(error.reason),
                url=url,
            ) from error
        except OSError as error:
            raise Md2HTMLAPIError(
                "Could not reach MD2HTML API: {}".format(error),
                url=url,
            ) from error

        text = raw.decode("utf-8", errors="replace")
        if status >= 400:
            raise Md2HTMLAPIError(
                "MD2HTML API returned HTTP {}".format(status),
                status_code=status,
                url=url,
                response_body=text,
            )
        if not text.strip():
            decoded: Any = None
        else:
            try:
                decoded = json.loads(text)
            except ValueError:
                decoded = text
        self.last_response = decoded
        return decoded

    @staticmethod
    def _require_text(value: Any, name: str) -> str:
        if not isinstance(value, str):
            raise TypeError("{} must be a string".format(name))
        return value

    @staticmethod
    def _require_object(response: Any, endpoint: str) -> Dict[str, Any]:
        if not isinstance(response, dict):
            raise Md2HTMLAPIError(
                "{} returned an unexpected response".format(endpoint),
                response_body=repr(response),
            )
        return response

    def convert(self, md: str) -> str:
        """Convert one Markdown document and return its HTML string."""
        md = self._require_text(md, "md")
        response = self._require_object(
            self._request("POST", "/convert", {"markdown": md}), "/convert"
        )
        if "html" not in response:
            raise Md2HTMLAPIError(
                "/convert response did not include 'html'",
                response_body=json.dumps(response),
            )
        return str(response["html"])

    def batch(self, items: Iterable[str]) -> List[str]:
        """Convert up to 50 Markdown documents and return HTML strings."""
        if isinstance(items, (str, bytes)):
            raise TypeError("items must be an iterable of Markdown strings")
        values = list(items)
        if not values:
            raise ValueError("items must contain at least one Markdown string")
        if len(values) > 50:
            raise ValueError("items cannot contain more than 50 documents")
        for item in values:
            self._require_text(item, "each item")

        response = self._request("POST", "/batch", {"items": values})
        if isinstance(response, list):
            return [str(item) for item in response]
        response_object = self._require_object(response, "/batch")
        results = response_object.get("results")
        if not isinstance(results, list):
            raise Md2HTMLAPIError(
                "/batch response did not include a results list",
                response_body=json.dumps(response_object),
            )
        return [str(item) for item in results]

    def register(self, email: str) -> Dict[str, Any]:
        """Register an email and return the API key response.

        Current hosted deployments historically exposed ``GET /register``
        without an email. The requested POST form is attempted first; a 404 or
        405 transparently falls back to that legacy endpoint.
        """
        email = self._require_text(email, "email")
        if not email.strip():
            raise ValueError("email must not be empty")
        try:
            response = self._request("POST", "/register", {"email": email})
        except Md2HTMLAPIError as error:
            if error.status_code not in (404, 405):
                raise
            response = self._request("GET", "/register")
        response_object = self._require_object(response, "/register")
        key = response_object.get("api_key")
        if isinstance(key, str) and key:
            self.api_key = key
        return response_object

    def get_usage(self) -> Dict[str, Any]:
        """Return usage and remaining free-tier calls."""
        return self._require_object(self._request("GET", "/usage"), "/usage")

    def prettify_json(self, data: Union[str, Dict[str, Any], List[Any]]) -> str:
        """Pretty-print JSON data and return the formatted JSON string."""
        if isinstance(data, str):
            json_text = data
        else:
            try:
                json_text = json.dumps(data, ensure_ascii=False)
            except (TypeError, ValueError) as error:
                raise TypeError("data must be JSON-serializable") from error
        response = self._request("POST", "/json/prettify", {"json": json_text})
        if isinstance(response, str):
            return response
        response_object = self._require_object(response, "/json/prettify")
        prettified = response_object.get("prettified")
        if isinstance(prettified, str):
            return prettified
        # A compatible self-hosted API may return the decoded JSON object.
        without_billing = dict(response_object)
        without_billing.pop("billing", None)
        return json.dumps(without_billing, indent=2, ensure_ascii=False)

    def text_stats(self, text: str) -> Dict[str, Any]:
        """Return word, character, reading-time, and top-word statistics."""
        text = self._require_text(text, "text")
        response = self._require_object(
            self._request("POST", "/text/stats", {"text": text}), "/text/stats"
        )
        result = dict(response)
        result.pop("billing", None)
        return result

    # Friendly aliases used by earlier versions of the SDK.
    usage = get_usage

    def __repr__(self) -> str:
        key = "set" if self.api_key else "unset"
        return "Md2HTMLClient(base_url={!r}, api_key={})".format(self.base_url, key)


# Backwards-compatible names for the first SDK draft.
MD2HTMLClient = Md2HTMLClient
MD2HTMLAPIError = Md2HTMLAPIError

__all__ = [
    "Md2HTMLClient",
    "Md2HTMLAPIError",
    "MD2HTMLClient",
    "MD2HTMLAPIError",
]

File: sdk/md2html-client.js
'use strict';

/**
 * MD2HTMLClient — Node.js SDK for the MD2HTML API.
 *
 * The MD2HTML API converts Markdown to HTML (and offers a few small text
 * utilities) behind a micropayment billing layer. This client wraps all 10
 * public endpoints with a thin fetch-based interface.
 *
 * Requirements: Node.js 18+ (global fetch). No external dependencies.
 *
 * Quick start:
 *   const { MD2HTMLClient } = require('./md2html-client');
 *   const client = new MD2HTMLClient({ apiKey: 'mk_...' });
 *   const { html } = await client.convert('# Hello **world**');
 *
 * All methods are async and return the parsed JSON response body. On a non-2xx
 * status they reject with an Error that carries `.status`, `.url`, and the
 * parsed response body (when available) under `.body`.
 */

const DEFAULT_BASE_URL = 'http://147.15.103.217/md2html';
const DEFAULT_TIMEOUT_MS = 30000;

class MD2HTMLClient {
  /**
   * @param {Object}  [opts]
   * @param {string}  [opts.baseUrl]   Base URL including the /md2html mount path.
   * @param {string}  [opts.apiKey]    API key (mk_...). Sent as X-API-Key. Optional.
   * @param {number}  [opts.timeoutMs] Request timeout in milliseconds.
   * @param {Object}  [opts.headers]   Extra default headers merged into every request.
   * @param {typeof fetch} [opts.fetch] Custom fetch implementation (testing).
   */
  constructor(opts = {}) {
    if (opts !== null && typeof opts !== 'object') {
      throw new TypeError('MD2HTMLClient: opts must be an object');
    }
    this.baseUrl = (opts.baseUrl || DEFAULT_BASE_URL).replace(/\/+$/, '');
    this.apiKey = opts.apiKey || null;
    this.timeoutMs = opts.timeoutMs || DEFAULT_TIMEOUT_MS;
    this.extraHeaders = opts.headers || {};
    this._fetch = opts.fetch || (typeof fetch !== 'undefined' ? fetch : null);
    if (!this._fetch) {
      throw new Error(
        'MD2HTMLClient: global fetch is unavailable. Use Node.js 18+ or pass opts.fetch.'
      );
    }
  }

  // ---- internals -----------------------------------------------------------

  /** Build an absolute URL from a path relative to baseUrl. */
  _url(path) {
    return `${this.baseUrl}${path}`;
  }

  /** Default headers for every request. */
  _defaultHeaders() {
    const h = Object.assign({}, this.extraHeaders);
    if (this.apiKey) h['X-API-Key'] = this.apiKey;
    return h;
  }

  /**
   * Core request helper. Always sends Accept: application/json and (for POST)
   * Content-Type: application/json unless overridden by callOptions.headers.
   * @private
   */
  async _request(method, path, { body, headers, signal } = {}) {
    const url = this._url(path);
    const reqHeaders = Object.assign(
      { Accept: 'application/json' },
      this._defaultHeaders(),
      headers || {}
    );

    const init = {
      method,
      headers: reqHeaders,
      signal,
    };

    if (body !== undefined) {
      if (typeof body === 'string') {
        init.body = body;
        if (!('Content-Type' in reqHeaders)) reqHeaders['Content-Type'] = 'text/plain';
      } else {
        init.body = JSON.stringify(body);
        if (!('Content-Type' in reqHeaders)) reqHeaders['Content-Type'] = 'application/json';
      }
    }

    // Enforce a timeout via AbortController (does not override caller's signal).
    let timeoutController;
    let timedOut = false;
    if (!signal) {
      timeoutController = new AbortController();
      init.signal = timeoutController.signal;
      setTimeout(() => {
        timedOut = true;
        timeoutController.abort();
      }, this.timeoutMs);
    }

    let res;
    try {
      res = await this._fetch(url, init);
    } catch (err) {
      if (timedOut || (err && err.name === 'AbortError')) {
        throw new Error(`MD2HTMLClient: request to ${url} timed out after ${this.timeoutMs}ms`);
      }
      throw new Error(`MD2HTMLClient: network error requesting ${url}: ${err.message}`);
    }

    const text = await res.text();
    let parsed;
    try {
      parsed = text ? JSON.parse(text) : null;
    } catch (_) {
      parsed = text;
    }

    if (!res.ok) {
      const e = new Error(
        `MD2HTMLClient: ${method} ${path} failed with status ${res.status}`
      );
      e.status = res.status;
      e.url = url;
      e.body = parsed;
      throw e;
    }
    return parsed;
  }

  // ---- public endpoints ----------------------------------------------------

  /**
   * POST /convert — Convert a Markdown string to HTML.
   * @param {string} markdown                Markdown source.
   * @param {Object} [opts]
   * @param {string} [opts.apiKey]            One-off API key override for this call.
   * @returns {Promise<{html:string, billing?:Object, warning?:string}>}
   */
  async convert(markdown, opts = {}) {
    if (typeof markdown !== 'string') {
      throw new TypeError('convert(markdown): expected a string');
    }
    const headers = {};
    if (opts.apiKey) headers['X-API-Key'] = opts.apiKey;
    return this._request('POST', '/convert', { body: { markdown }, headers });
  }

  /**
   * POST /json/prettify — Re-indent a compact JSON string with 2-space pretty printing.
   * @param {string} json  A compact JSON string, e.g. '{"a":1,"b":2}'.
   * @returns {Promise<Object>} Prettified JSON string plus a billing object.
   */
  async jsonPrettify(json) {
    if (typeof json !== 'string') {
      throw new TypeError('jsonPrettify(json): expected a JSON string');
    }
    return this._request('POST', '/json/prettify', { body: { json } });
  }

  /**
   * POST /text/stats — Word/char counts, reading time, and top words for a text.
   * @param {string} text
   * @returns {Promise<{words:number, chars:number, chars_no_spaces:number,
   *                     reading_time_min:number, top_words:Array, billing?:Object}>}
   */
  async textStats(text) {
    if (typeof text !== 'string') {
      throw new TypeError('textStats(text): expected a string');
    }
    return this._request('POST', '/text/stats', { body: { text } });
  }

  /**
   * POST /slug — Generate a URL-safe slug from a title.
   * @param {string} title
   * @returns {Promise<{slug:string, billing?:Object}>}
   */
  async slug(title) {
    if (typeof title !== 'string') {
      throw new TypeError('slug(title): expected a string');
    }
    return this._request('POST', '/slug', { body: { title } });
  }

  /**
   * GET /register — Mint a new API key and its associated wallet address.
   * @returns {Promise<{api_key:string, wallet_address:string,
   *                     free_tier_limit:number, calls_made:number, remaining:number}>}
   */
  async register() {
    return this._request('GET', '/register');
  }

  /**
   * GET /health — Liveness probe. Returns service uptime and status JSON. No auth.
   * @returns {Promise<{status:string, version?:string, uptime_seconds:number, port:number}>}
   */
  async health() {
    return this._request('GET', '/health');
  }

  /**
   * GET /docs — Interactive API documentation (auto-generated).
   * @returns {Promise<string|Object>} The docs page body (text or JSON depending on server).
   */
  async docs() {
    return this._request('GET', '/docs', { headers: { Accept: 'text/plain, application/json' } });
  }

  /**
   * GET /payment — Returns the LTC wallet address for topping up quota.
   * @returns {Promise<{wallet_address:string, currency:string}>}
   */
  async payment() {
    return this._request('GET', '/payment');
  }

  /** POST /payment/claim — Verify a confirmed LTC txid and add prepaid calls. */
  async claimPayment(txid, opts = {}) {
    if (typeof txid !== 'string' || !/^[0-9a-fA-F]{64}$/.test(txid.trim())) {
      throw new TypeError('claimPayment(txid): expected 64 hexadecimal characters');
    }
    const apiKey = opts.apiKey || this.apiKey;
    if (!apiKey) {
      throw new Error('claimPayment(txid): an API key is required');
    }
    return this._request('POST', '/payment/claim', {
      body: { txid: txid.trim() },
      headers: { 'X-API-Key': apiKey },
    });
  }

  /**
   * GET /usage — Current quota usage for the caller (API key or IP).
   * @param {Object} [opts]
   * @param {string} [opts.apiKey] One-off API key override.
   * @returns {Promise<{calls_made:number, remaining:number}>}
   */
  async usage(opts = {}) {
    const headers = {};
    if (opts.apiKey) headers['X-API-Key'] = opts.apiKey;
    return this._request('GET', '/usage', { headers });
  }

  /**
   * GET /stats — Aggregate, public, read-only service metrics.
   * @returns {Promise<{total_calls:number, unique_ips:number}>}
   */
  async stats() {
    return this._request('GET', '/stats');
  }
}

// ---- CommonJS + ESM dual export ------------------------------------------

module.exports = { MD2HTMLClient };
module.exports.MD2HTMLClient = MD2HTMLClient;
module.exports.default = MD2HTMLClient;

// `Object.defineProperty` makes default-ESM-import interop work without
// breaking `require('./md2html-client').MD2HTMLClient`.
if (typeof Symbol !== 'undefined' && Symbol.toStringTag) {
  MD2HTMLClient.prototype[Symbol.toStringTag] = 'MD2HTMLClient';
}
